src/Nethermind/Nethermind.Core/ContainerBuilderExtensions.cs

```diff
@@ -81,22 +72,16 @@ public static ContainerBuilder AddSingleton<T, TArg0, TArg1>(this ContainerBuild
 
     public static ContainerBuilder AddSingleton<T, TArg0, TArg1, TArg2>(this ContainerBuilder builder, Func<TArg0, TArg1, TArg2, T> factoryMethod) where T : class where TArg0 : notnull where TArg1 : notnull where TArg2 : notnull
     {
-        builder.Register((ctx) =>
-            {
-                MethodInfo factoryMethodInfo = factoryMethod.Method;
-
-                TArg0 arg0 = factoryMethodInfo.GetParameters()[0].GetCustomAttribute<KeyFilterAttribute>() is { } keyFilter
-                    ? ctx.ResolveKeyed<TArg0>(keyFilter.Key)
-                    : ctx.Resolve<TArg0>();
-                TArg1 arg1 = factoryMethodInfo.GetParameters()[1].GetCustomAttribute<KeyFilterAttribute>() is { } keyFilter1
-                    ? ctx.ResolveKeyed<TArg1>(keyFilter1.Key)
-                    : ctx.Resolve<TArg1>();
-                TArg2 arg2 = factoryMethodInfo.GetParameters()[2].GetCustomAttribute<KeyFilterAttribute>() is { } keyFilter2
-                    ? ctx.ResolveKeyed<TArg2>(keyFilter2.Key)
-                    : ctx.Resolve<TArg2>();
-
-                return factoryMethod(arg0, arg1, arg2);
-            })
+        Func<IComponentContext, TArg0> param0 = CreateArgResolver<TArg0>(factoryMethod.Method, 0);
+        Func<IComponentContext, TArg1> param1 = CreateArgResolver<TArg1>(factoryMethod.Method, 1);
+        Func<IComponentContext, TArg2> param2 = CreateArgResolver<TArg2>(factoryMethod.Method, 2);
+
+        builder
+            .Register((ctx) => factoryMethod(
+                param0(ctx),
+                param1(ctx),
+                param2(ctx)
+            ))
```

Comment:
could even be in-lined:

```suggestion
        builder
            .Register((ctx) => factoryMethod(
                CreateArgResolver<TArg0>(factoryMethod.Method, 0)(ctx),
                CreateArgResolver<TArg1>(factoryMethod.Method, 1(ctx),
                CreateArgResolver<TArg2>(factoryMethod.Method, 2)(ctx)
            ))
```